contracts/cosmwasm-vm/cw-cluster-connection/src/contract.rs

```diff
@@ -0,0 +1,302 @@
+        let new_admin = deps.api.addr_validate(address.as_str())?;
+        let _ = self.store_admin(deps.storage, new_admin);
+        Ok(Response::new().add_attribute("action", "set_admin"))
+    }
+
+    pub fn set_relayer(
+        &mut self,
+        deps: DepsMut,
+        info: MessageInfo,
+        address: Addr,
+    ) -> Result<Response, ContractError> {
+        self.ensure_admin(deps.storage, info.sender)?;
+        let new_relayer = deps.api.addr_validate(address.as_str())?;
+        let _ = self.store_relayer(deps.storage, new_relayer);
+        Ok(Response::new().add_attribute("action", "set_relayer"))
+    }
+
+    pub fn set_validators(
+        &mut self,
+        deps: DepsMut,
+        info: MessageInfo,
+        validators: Vec<String>,
+        threshold: u8,
+    ) -> Result<Response, ContractError> {
+        self.ensure_admin(deps.storage, info.sender)?;
+
+        if threshold < 1 {
+            return Err(ContractError::InvalidThreshold {
+                msg: "threshold should be at least 1".to_string(),
+            });
+        }
+
+        if validators.len() > 0 {
+            self.clear_validators(deps.storage)?;
+            for rlr in validators {
+                self.store_validator(deps.storage, rlr)?;
+            }
+        }
+
+        let validators_set = self.get_validators(deps.storage)?;
+
+        if threshold as usize > validators_set.len() {
+            return Err(ContractError::InvalidThreshold {
+                msg: "threshold should be at most the size of validators".to_string(),
+            });
+        }
+
+        self.store_signature_threshold(deps.storage, threshold)?;
+
+        Ok(Response::new().add_attribute("action", "set_validators"))
+    }
+
+    pub fn send_message(
+        &mut self,
+        deps: DepsMut,
+        info: MessageInfo,
+        to: NetId,
+        sn: i64,
+        msg: Vec<u8>,
+    ) -> Result<Response, ContractError> {
+        self.ensure_xcall(deps.storage, info.sender)?;
+
+        let next_conn_sn = self.get_next_conn_sn(deps.storage)?;
+
+        let mut fee = 0;
+
+        if sn >= 0 {
+            fee = self.get_fee(deps.storage, to.clone(), sn > 0)?.into();
+        }
+
+        let value = self.get_amount_for_denom(&info.funds, self.get_denom(deps.storage));
+
+        if fee > value {
+            return Err(ContractError::InsufficientFunds);
+        }
+
+        Ok(Response::new()
+            .add_attribute("action", "send_message")
+            .add_event(
+                Event::new("Message")
+                    .add_attribute("targetNetwork", to.to_string())
+                    .add_attribute("connSn", next_conn_sn.to_string())
+                    .add_attribute("msg", self.hex_encode(msg)),
+            ))
+    }
+
+    pub fn recv_message(
+        &mut self,
+        deps: DepsMut,
+        info: MessageInfo,
+        src_network: NetId,
+        conn_sn: u128,
+        msg: String,
+    ) -> Result<Response, ContractError> {
```

Comment:
where are signatures here? having this method will allow relayer to bypass signature check why are we having this? I saw this in another contract as well.

Reply:
now there is